homeassistant/components/firmata/__init__.py

```diff
@@ -0,0 +1,142 @@
+    },
+    required=True,
+    extra=vol.ALLOW_EXTRA,
+)
+
+BOARD_CONFIG_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_SERIAL_PORT): cv.string,
+        vol.Optional(CONF_SERIAL_BAUD_RATE): cv.positive_int,
+        vol.Optional(CONF_ARDUINO_INSTANCE_ID): cv.positive_int,
+        vol.Optional(CONF_ARDUINO_WAIT): cv.positive_int,
+        vol.Optional(CONF_SLEEP_TUNE): vol.All(
+            vol.Coerce(float), vol.Range(min=0.0001)
+        ),
+        vol.Optional(CONF_SAMPLING_INTERVAL): cv.positive_int,
+        vol.Optional(CONF_SWITCHES): [SWITCH_SCHEMA],
+        vol.Optional(CONF_BINARY_SENSORS): [BINARY_SENSOR_SCHEMA],
+    },
+    required=True,
+    extra=vol.ALLOW_EXTRA,
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.All(cv.ensure_list, [BOARD_CONFIG_SCHEMA])}, extra=vol.ALLOW_EXTRA
+)
+
+
+async def async_setup(hass, config):
+    """Set up the Firmata domain."""
+    # Delete all current entries and re-add them to update them
+    if hass.config_entries.async_entries(DOMAIN) and DOMAIN in config:
+        for entry in hass.config_entries.async_entries(DOMAIN):
+            await hass.config_entries.async_remove(entry.entry_id)
+    if DOMAIN in config:
+        for board in config[DOMAIN]:
+            firmata_config = deepcopy(board)
+            hass.async_create_task(
+                hass.config_entries.flow.async_init(
+                    DOMAIN,
+                    context={"source": config_entries.SOURCE_IMPORT},
+                    data=firmata_config,
+                )
+            )
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set up a Firmata board for a config entry."""
+    if DOMAIN not in hass.data:
+        hass.data[DOMAIN] = {}
+
+    _LOGGER.debug(
+        "Setting up Firmata id %s, config %s", config_entry.entry_id, config_entry.data
+    )
+    if not config_entry.options:
+        hass.config_entries.async_update_entry(config_entry)
+
+    board = FirmataBoard(hass, config_entry)
+
+    if not await board.async_setup():
+        return False
+
+    hass.data[DOMAIN][board.name] = board
+
+    await board.async_update_device_registry()
+    if CONF_BINARY_SENSORS in config_entry.data:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, "binary_sensor")
+        )
+    if CONF_SWITCHES in config_entry.data:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, "switch")
+        )
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, board.async_reset)
+    return True
+
+
+async def async_remove_entry(hass, config_entry) -> None:
+    """Shutdown and close a Firmata board for a config entry."""
+    name = config_entry.data[CONF_NAME]
+
+    if DOMAIN in hass.data:
+        if name in hass.data[DOMAIN]:
+            _LOGGER.info("Closing Firmata board %s", name)
+            await hass.data[DOMAIN][name].async_reset()
```

Comment:
You should remove the board from `hass.data[DOMAIN]`
```suggestion
            await hass.data[DOMAIN].pop(name).async_reset()
```